### PR TITLE
JoinWithTable

### PR DESCRIPTION
**Pull request by talIguaz**

### CODE REVIEW

File: integration/integration_test_utils.py [hunk 1]
```diff
@@ -85,6 +90,30 @@ def setup_teardown_test():
...
+    asyncio.run(create_temp_kv(table_name))
+
+    # Test runs
+    yield table_name
+
+    # Teardown
+    asyncio.run(recursive_delete(table_name, V3ioHeaders()))
+
+
+async def create_temp_kv(table_path):
```
Comment (gtopper) [inline integration/integration_test_utils.py:106]: Is it table name or table path?

File: storey/flow.py [hunk 2]
```diff
@@ -651,6 +651,50 @@ async def _cleanup(self):
...
 
 
+class JoinWithTable(_ConcurrentJobExecution):
+    """Joins each event with data from the given table.
+
+    :param table: Table to join with.
+    :type table: Table
+    :param key_extractor: Function for extracting the key for table access from an event.
+    :type key_extractor: Function (Event=>string)
+    :param attributes: A comma-separated list of attributes to be requested from V3IO. Defaults to '*' (all user attributes).
```
Comment (gtopper) [inline storey/flow.py]: V3IO?

File: storey/flow.py [hunk 3]
```diff
@@ -651,6 +651,50 @@ async def _cleanup(self):
...
+    :type attributes: list of string
```
Comment (gtopper) [inline storey/flow.py:662]: list of string defualt to `'*'`?

File: storey/flow.py [hunk 4]
```diff
@@ -651,6 +651,50 @@ async def _cleanup(self):
...
+    :param join_function: Joins the original event with relevant data received from V3IO.
```
Comment (gtopper) [inline storey/flow.py]: V3IO?

Comment (gtopper) [reply storey/flow.py]: Need to mention the default behaviour.

File: storey/flow.py [hunk 5]
```diff
@@ -651,6 +651,50 @@ async def _cleanup(self):
...
+
+    def __init__(self, table, key_extractor, attributes='*', join_function=None, **kwargs):
+        super().__init__(**kwargs)
+
+        self._table = table
+        self._closeables = [table]
+
+        self._key_extractor = key_extractor
+
+        def _default_join_fn(event, join_res):
```
Comment (gtopper) [inline storey/flow.py]: No need to protect local scope with an underscore.

File: storey/drivers.py [hunk 6]
```diff
@@ -365,16 +365,16 @@ async def _load_aggregates_by_key(self, container, table_path, key):
...
 
-        response = await self._v3io_client.kv.get(container, table_path, key, raise_for_status=v3io.aio.dataplane.RaiseForStatus.never)
+        response = await self._get_item(container, table_path, key, attributes)
         if response.status_code == 404:
             return None
         elif response.status_code == 200:
             res = {}
             for name in response.output.item.keys():
-                if not name.startswith(self._aggregation_attribute_prefix):
+                if not name.startswith((self._aggregation_attribute_prefix, self._aggregation_time_attribute_prefix)):
```
Comment (gtopper) [inline storey/drivers.py:377]: ❓

File: integration/test_flow_integration.py [hunk 7]
```diff
@@ -168,3 +168,39 @@ def test_write_to_tsdb():
     df = pd.DataFrame(expected, columns=columns)
     df.set_index(keys=['time'], inplace=True)
     assert res.equals(df), f"result{res}\n!=\nexpected{df}"
+
+
+def test_join_by_key(setup_kv_teardown_test):
+    table = Table(setup_kv_teardown_test, V3ioDriver())
```
Comment (gtopper) [inline integration/test_flow_integration.py:174]: Can we have a unit test as well? Probably need to add a static data table for that.

File: storey/drivers.py (hunk 6 above)
Comment (talIguaz) [reply storey/drivers.py:377]: There are 2 different system prefixes, and im ignoring both here

Comment (gtopper) [reply storey/drivers.py:377]: Huh, I didn't know you could pass a tuple there.

**Review by gtopper — APPROVED** [review]

**Pull request merged**